src/mmds/src/token.rs

```diff
@@ -0,0 +1,563 @@
+        let encoded_token = token.base64_encode_token()?;
+        // Increase the count of encrypted tokens.
+        self.num_encrypted_tokens += 1;
+
+        Ok(encoded_token)
+    }
+
+    /// Create a new Token structure to encrypt.
+    fn create_token(&mut self, ttl_seconds: u32) -> Result<Token, Error> {
+        // Validate token time to live against bounds.
+        if !TokenAuthority::check_ttl(ttl_seconds) {
+            return Err(Error::InvalidTtlValue(ttl_seconds));
+        }
+
+        // Generate 12-byte random nonce.
+        let mut iv = [0u8; IV_LEN];
+        self.entropy_pool
+            .read_exact(&mut iv)
+            .map_err(|_| Error::TokenGeneration)?;
+
+        // Compute expiration time from ttl.
+        let expiry = TokenAuthority::compute_expiry(ttl_seconds);
+        // Encrypt expiry using the nonce.
+        let (payload, tag) = self.encrypt_expiry(expiry, iv.as_ref())?;
+
+        Ok(Token::new(iv, payload, tag))
+    }
+
+    /// Encrypt expiry using AES-GCM block cipher and return payload and tag obtained.
+    fn encrypt_expiry(
+        &self,
+        expiry: u64,
+        iv: &[u8],
+    ) -> Result<([u8; PAYLOAD_LEN], [u8; TAG_LEN]), Error> {
+        // Create Nonce object from initialization vector.
+        let nonce = Nonce::from_slice(iv);
+        // Convert expiry u64 value into bytes.
+        let mut expiry_as_bytes = expiry.to_le_bytes();
+
+        let tag = self
+            .cipher
+            .encrypt_in_place_detached(nonce, b"", &mut expiry_as_bytes)
+            .map_err(|_| Error::TokenGeneration)?;
+
+        // Tag must be of size `TAG_LEN`.
+        let tag_as_bytes: [u8; TAG_LEN] = tag
+            .as_slice()
+            .try_into()
+            .map_err(|_| Error::TokenGeneration)?;
+
+        Ok((expiry_as_bytes, tag_as_bytes))
+    }
+
+    /// Attempts to decrypt expiry value within token sequence. Returns false if expiry
+    /// cannot be decrypted. If decryption succeeds, returns true if token has not expired
+    /// (i.e. current time is greater than expiry) and false otherwise.
+    pub fn is_valid(&self, encoded_token: &str) -> bool {
+        // Check size of encoded token struct.
+        if encoded_token.len() > TOKEN_LENGTH_LIMIT {
+            return false;
+        }
+
+        // Decode token struct from base64.
+        let mut token = match Token::base64_decode_token(encoded_token) {
+            Ok(token) => token,
+            Err(_) => return false,
+        };
+
+        // Validate constant fields inside decoded token structure.
+        if !token.sanity_check() {
+            return false;
+        }
+
+        // Decrypt ttl using AES-GCM block cipher.
+        let expiry = match self.decrypt_expiry(&mut token.payload, &token.tag, &token.iv) {
+            Ok(expiry) => expiry,
+            Err(_) => return false,
+        };
+
+        // Get current time.
+        let time_now = SystemTime::now()
```

Comment:
This uses the realtime clock of the system, a clocksource that be reconfigured by other users in the system and that may be affected by drifts (both forward and backward).

Since we are interested in the absolute difference here, I suggest we use a monotonic clock. This is a clock type that is initialised when the system is booted and is guaranteed to only go forwards and never be adjusted by external actors.

There are helpers in `utils/src/time.rs` for working with clocks.

Reply:
Looks good 👍🏻